### ISSUE TITLE
Issue with Flux Training, Tensor Size Mismatch

### ISSUE THREAD

**Issue by EthicalCell** [tdrussell/diffusion-pipe#45]
```
Flux .toml config: 

# Output path for training runs. Each training run makes a new directory in here.
output_dir = '/workspace/ComfyUI/models/loras/flux'

# Dataset config file.
dataset = 'examples/dataset.toml'
# You can have separate eval datasets. Give them a name for Tensorboard metrics.
# eval_datasets = [
#     {name = 'something', config = 'path/to/eval_dataset.toml'},
# ]

# training settings

# I usually set this to a really high value because I don't know how long I want to train.
epochs = 1000
# Batch size of a single forward/backward pass for one GPU.
micro_batch_size_per_gpu = 1
# Pipeline parallelism degree. A single instance of the model is divided across this many GPUs.
pipeline_stages = 1
# Number of micro-batches sent through the pipeline for each training step.
# If pipeline_stages > 1, a higher GAS means better GPU utilization due to smaller pipeline bubbles (where GPUs aren't overlapping computation).
gradient_accumulation_steps = 4
# Grad norm clipping.
gradient_clipping = 1.0
# Learning rate warmup.
warmup_steps = 100

# eval settings

eval_every_n_epochs = 1
eval_before_first_step = true
# Might want to set these lower for eval so that less images get dropped (eval dataset size is usually much smaller than training set).
# Each size bucket of images/videos is rounded down to the nearest multiple of the global batch size, so higher global batch size means
# more dropped images. Usually doesn't matter for training but the eval set is much smaller so it can matter.
eval_micro_batch_size_per_gpu = 1
eval_gradient_accumulation_steps = 1

# misc settings

# Probably want to set this a bit higher if you have a smaller dataset so you don't end up with a million saved models.
save_every_n_epochs = 1
# Can checkpoint the training state every n number of epochs or minutes. Set only one of these. You can resume from checkpoints using the --resume_from_checkpoint flag.
#checkpoint_every_n_epochs = 1
checkpoint_every_n_minutes = 120
# Always set to true unless you have a huge amount of VRAM.
activation_checkpointing = true
# Controls how Deepspeed decides how to divide layers across GPUs. Probably don't change this.
partition_method = 'parameters'
# dtype for saving the LoRA or model, if different from training dtype
save_dtype = 'bfloat16'
# Batch size for caching latents and text embeddings. Increasing can lead to higher GPU utilization during caching phase but uses more memory.
caching_batch_size = 1
# How often deepspeed logs to console.
steps_per_print = 1
# How to extract video clips for training from a single input video file.
# The video file is first assigned to one of the configured frame buckets, but then we must extract one or more clips of exactly the right
# number of frames for that bucket.
# single_beginning: one clip starting at the beginning of the video
# single_middle: one clip from the middle of the video (cutting off the start and end equally)
# multiple_overlapping: extract the minimum number of clips to cover the full range of the video. They might overlap some.
# default is single_middle
video_clip_mode = 'single_middle'

# [model]
# # flux, ltx-video, or hunyuan-video
# type = 'hunyuan-video'
# # Can load Hunyuan Video entirely from the ckpt path set up for the official inference scripts.
# #ckpt_path = '/home/anon/HunyuanVideo/ckpts'
# # Or you can load it by pointing to all the ComfyUI files.
# transformer_path = '/data2/imagegen_models/hunyuan_video_comfyui/hunyuan_video_720_cfgdistill_fp8_e4m3fn.safetensors'
# vae_path = '/data2/imagegen_models/hunyuan_video_comfyui/hunyuan_video_vae_bf16.safetensors'
# llm_path = '/data2/imagegen_models/hunyuan_video_comfyui/llava-llama-3-8b-text-encoder-tokenizer'
# clip_path = '/data2/imagegen_models/hunyuan_video_comfyui/clip-vit-large-patch14'
# # Base dtype used for all models.
# dtype = 'bfloat16'
# # Hunyuan Video supports fp8 for the transformer when training LoRA.
# transformer_dtype = 'float8'
# # How to sample timesteps to train on. Can be logit_normal or uniform.
# timestep_sample_method = 'logit_normal'

# flux example
[model]
type = 'flux'
# Path to Huggingface Diffusers directory for Flux
diffusers_path = '/workspace/diffusion-pipe/FLUX.1-dev'
# You can override the transformer from a BFL format checkpoint.
transformer = '/workspace/diffusion-pipe/FLUX.1-dev/flux1-dev.safetensors'
dtype = 'bfloat16'
flux_shift = true

# LTV-Video example
# [model]
# type = 'ltx-video'
# diffusers_path = '/data2/imagegen_models/LTX-Video'
# dtype = 'bfloat16'
# timestep_sample_method = 'logit_normal'

[adapter]
type = 'lora'
rank = 32
# Dtype for the LoRA weights you are training.
dtype = 'bfloat16'
# You can initialize the lora weights from a previously trained lora.
#init_from_existing = '/data/diffusion_pipe_training_runs/something/epoch50'

[optimizer]
# AdamW from the optimi library is a good default since it automatically uses Kahan summation when training bfloat16 weights.
# Look at train.py for other options. You could also easily edit the file and add your own.
type = 'adamw8bit'
lr = 1e-4
betas = [0.9, 0.99]
weight_decay = 0.01
eps = 1e-8

_________________________________________________________________________________________________

Error Log:
[2024-12-28 14:23:45,277] [INFO] [config.py:1003:print]   zero_config .................. stage=0 contiguous_gradients=True reduce_scatter=True reduce_bucket_size=500000000 use_multi_rank_bucket_allreduce=True allgather_partitions=True allgather_bucket_size=500000000 overlap_comm=False load_from_fp32_weights=True elastic_checkpoint=False offload_param=None offload_optimizer=None sub_group_size=1000000000 cpu_offload_param=None cpu_offload_use_pin_memory=None cpu_offload=None prefetch_bucket_size=50000000 param_persistence_threshold=100000 model_persistence_threshold=9223372036854775807 max_live_parameters=1000000000 max_reuse_distance=1000000000 gather_16bit_weights_on_model_save=False module_granularity_threshold=0 use_all_reduce_for_fetch_params=False stage3_gather_fp16_weights_on_model_save=False ignore_unused_parameters=True legacy_stage1=False round_robin_gradients=False zero_hpz_partition_size=1 zero_quantized_weights=False zero_quantized_nontrainable_weights=False zero_quantized_gradients=False zeropp_loco_param=None mics_shard_size=-1 mics_hierarchical_params_gather=False memory_efficient_linear=True pipeline_loading_checkpoint=False override_module_apply=True
[2024-12-28 14:23:45,277] [INFO] [config.py:1003:print]   zero_enabled ................. False
[2024-12-28 14:23:45,277] [INFO] [config.py:1003:print]   zero_force_ds_cpu_optimizer .. True
[2024-12-28 14:23:45,277] [INFO] [config.py:1003:print]   zero_optimization_stage ...... 0
[2024-12-28 14:23:45,277] [INFO] [config.py:989:print_user_config]   json = {
    "train_micro_batch_size_per_gpu": 1, 
    "gradient_accumulation_steps": 4, 
    "gradient_clipping": 1.0, 
    "steps_per_print": 1
}
[2024-12-28 14:23:45,277] [INFO] [engine.py:105:__init__] CONFIG: micro_batches=4 micro_batch_size=1
[2024-12-28 14:23:45,277] [INFO] [engine.py:146:__init__] is_pipe_partitioned= False is_grad_partitioned= False
[2024-12-28 14:23:45,324] [INFO] [engine.py:165:__init__] RANK=0 STAGE=0 LAYERS=60 [0, 60) STAGE_PARAMS=171935744 (171.936M) TOTAL_PARAMS=171935744 (171.936M) UNIQUE_PARAMS=171935744 (171.936M)
[rank0]: Traceback (most recent call last):
[rank0]:   File "/workspace/diffusion-pipe/train.py", line 458, in <module>
[rank0]:     loss = model_engine.train_batch().item()
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/pipe/engine.py", line 389, in train_batch
[rank0]:     self._exec_schedule(sched)
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/pipe/engine.py", line 1422, in _exec_schedule
[rank0]:     self._exec_instr(**cmd.kwargs)
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/pipe/engine.py", line 735, in _exec_forward_pass
[rank0]:     outputs = super().forward(inputs)
[rank0]:               ^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/utils/nvtx.py", line 18, in wrapped_fn
[rank0]:     ret_val = func(*args, **kwargs)
[rank0]:               ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/engine.py", line 1914, in forward
[rank0]:     loss = self.module(*inputs, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/pipe/module.py", line 386, in forward
[rank0]:     x = self.activation_checkpoint_func(exec_range_func(start_idx, end_idx), *x)
[rank0]:         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 600, in _fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/activation_checkpointing/checkpointing.py", line 953, in checkpoint
[rank0]:     CheckpointFunction.apply(function, all_outputs, *args)
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/autograd/function.py", line 574, in apply
[rank0]:     return super().apply(*args, **kwargs)  # type: ignore[misc]
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/activation_checkpointing/checkpointing.py", line 544, in forward
[rank0]:     outputs = run_function(*inputs_cuda)
[rank0]:               ^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/deepspeed/runtime/pipe/module.py", line 363, in exec_func
[rank0]:     inputs = layer(inputs)
[rank0]:              ^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/amp/autocast_mode.py", line 43, in decorate_autocast
[rank0]:     return func(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/models/flux.py", line 321, in forward
[rank0]:     encoder_hidden_states, hidden_states = self.block(
[rank0]:                                            ^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/diffusers/models/transformers/transformer_flux.py", line 180, in forward
[rank0]:     attention_outputs = self.attn(
[rank0]:                         ^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/diffusers/models/attention_processor.py", line 588, in forward
[rank0]:     return self.processor(
[rank0]:            ^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/diffusers/models/attention_processor.py", line 2318, in __call__
[rank0]:     query = apply_rotary_emb(query, image_rotary_emb)
[rank0]:             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/workspace/diffusion-pipe/venv/lib/python3.11/site-packages/diffusers/models/embeddings.py", line 1208, in apply_rotary_emb
[rank0]:     out = (x.float() * cos + x_rotated.float() * sin).to(x.dtype)
[rank0]:            ~~~~~~~~~~^~~~~
[rank0]: RuntimeError: The size of tensor a (4608) must match the size of tensor b (16896) at non-singleton dimension 2
[2024-12-28 14:23:58,874] [INFO] [launch.py:319:sigkill_handler] Killing subprocess 7470
[2024-12-28 14:23:58,875] [ERROR] [launch.py:325:sigkill_handler] ['/workspace/diffusion-pipe/venv/bin/python', '-u', 'train.py', '--local_rank=0', '--deepspeed', '--config', 'examples/hunyuan_video.toml'] exits with return code = 1

**Comment by tdrussell** [comment]
Should be fixed by 5b9c0be.

**Comment by EthicalCell** [comment]
Confirmed it's working, thanks! Amazing tool, appreciate all your work.
